README.md
# Sleeper



history
https://style.nikkei.com/article/DGXMZO24812560Z11C17A2000000?channel=DF260120166530
http://news.nicovideo.jp/watch/nw3270113
http://business.nikkeibp.co.jp/atcl/skillup/16/091500011/020200019/
https://prtimes.jp/main/html/rd/p/000000065.000019730.html
http://www.agingstyle.com/2018/02/07002386.html
http://wol.nikkeibp.co.jp/atcl/column/15/011500044/011600136/
https://www.mylohas.net/2018/02/aromarich_4.html
https://ddnavi.com/interview/435485/a/
https://japan.cnet.com/article/35114645/
https://fabcross.jp/news/2018/20180215_smartsleepmask_dreamlight.html
https://www.jscore.co.jp/column/lifestyle/2018/sleeper/
https://news.nifty.com/article/item/neta/12161-moco-43681/
https://prtimes.jp/main/html/rd/p/000000226.000005157.html
http://diamond.jp/articles/-/159818
https://news.mynavi.jp/article/20180221-586756/
http://top.tsite.jp/news/woman/o/38814470/
https://news.nifty.com/article/item/neta/12101-27131/
https://www.businessinsider.jp/post-161308
https://news.nifty.com/article/domestic/society/12184-38396/
https://www.bcnretail.com/market/detail/20180226_52725.html
https://news.mynavi.jp/article/20180225-589510/
https://jisin.jp/serial/%E3%82%A8%E3%83%B3%E3%82%BF%E3%83%A1/usako/33100
https://www.zaikei.co.jp/article/20180301/428931.html
http://www.elle.co.jp/decor/pick/sleep_18_0223
http://business.nikkeibp.co.jp/atcl/skillup/16/091500011/022300020/
http://top.tsite.jp/news/pet/o/39103779/
https://response.jp/article/2018/03/05/306783.html
https://getnavi.jp/healthcare/231769/
https://www.nikkansports.com/soccer/world/news/201803070000642.html
http://natgeo.nikkeibp.co.jp/atcl/web/15/403964/030600082/
https://kumanichi.com/kumacole/medical/377569/
https://gigazine.net/news/20180309-making-love-sleep/
https://www.nishinippon.co.jp/feature/life_topics/article/400840
https://prtimes.jp/main/html/rd/p/000000070.000018482.html
https://style.nikkei.com/article/DGXMZO28348140Q8A320C1000000?channel=DF260120166530
https://style.nikkei.com/article/DGXMZO28348140Q8A320C1000000?n_cid=LMNST020
http://toyokeizai.net/articles/-/213888
https://select.mamastar.jp/240876
https://hbol.jp/162195
https://www.bcnretail.com/market/detail/20180328_55671.html
https://news.biglobe.ne.jp/domestic/0329/ym_180329_9404511004.html
https://mainichi.jp/articles/20180404/k00/00m/020/054000c
https://news.biglobe.ne.jp/trend/0402/mai_180402_7133573936.html
http://news.nicovideo.jp/watch/nw3412212
https://toyokeizai.net/articles/-/213888
https://www.nikkei.com/article/DGXMZO28675940Y8A320C1000000/
http://diamond.jp/articles/-/165672
https://www.nishinippon.co.jp/nnp/industrial_physician/article/407251
https://getnavi.jp/digital/246355/
https://japan.cnet.com/article/35117444/
https://www.sankei.com/life/news/180410/lif1804100022-n1.html
https://info.ninchisho.net/archives/28828
https://www.nishinippon.co.jp/nnp/industrial_physician/article/407442
https://mainichi.jp/premier/health/articles/20180403/med/00m/010/014000c
http://ascii.jp/elem/000/001/664/1664580/
http://www.elle.co.jp/decor/news/nokia18-0417
https://www.rodo.co.jp/news/43156/
https://dime.jp/genre/536759/
http://natgeo.nikkeibp.co.jp/atcl/web/15/403964/041800085/
http://president.jp/articles/-/24711
https://style.nikkei.com/article/DGXKZO28734210Z20C18A3W10600?channel=DF130120166090
https://response.jp/article/2018/04/23/308852.html
https://japanese.engadget.com/2018/04/23/nokia-sleep-ifttt-iot/
https://style.nikkei.com/article/DGXKZO28734210Z20C18A3W10600?channel=DF140920160919
https://getnavi.jp/healthcare/254228/
https://dime.jp/genre/540736/
https://www.value-press.com/pressrelease/200648
http://healthpress.jp/2018/04/post-3602.html
https://nikkan-spa.jp/1470890
https://news.biglobe.ne.jp/trend/0501/aab_180501_1743982899.html
https://www.asahi.com/articles/ASL4W5G0LL4WPTIL038.html
https://woman.excite.co.jp/article/lifestyle/rid_Folk_942425/
https://www.nikkei.com/article/DGXMZO30015060R00C18A5I10000/
https://www.j-cast.com/bookwatch/2018/05/05007307.html
https://news.nifty.com/article/item/neta/12101-28806/
https://www.mylohas.net/2018/05/166867sleep.html
https://newswitch.jp/p/12876
https://www.fnn.jp/posts/00391628CX
https://news.mynavi.jp/article/20180509-628428/
https://www.nikkei.com/article/DGXMZO29215330Q8A410C1000000/
https://www.nikkei.com/article/DGXMZO3037533011052018000000/
https://ananweb.jp/anew/148185/
https://www.roomie.jp/2018/05/429468/
https://www.nikkei.com/article/DGXMZO30408690R10C18A5000000/
https://motor-fan.jp/tech/10004159
https://www.roomie.jp/2018/05/430023/
https://scienceportal.jst.go.jp/news/newsflash_review/newsflash/2018/05/20180515_01.html
https://news.yahoo.co.jp/byline/yatsuzukaeri/20180515-00085245/
https://yomidr.yomiuri.co.jp/article/20180517-OYTEW259122/
http://www.mag2.com/p/news/358962
https://toyokeizai.net/articles/-/219127
http://blogos.com/article/298997/
https://www.fnn.jp/posts/00313960HDK
https://www.asahi.com/articles/ASL4W5FTML4WUTIL02J.html
https://www.cnn.co.jp/fringe/35119783.html
http://healthpress.jp/2018/05/post-3643.html
https://www.asahi.com/articles/ASL5X5GMKL5XUTFK01M.html
https://toyokeizai.net/articles/-/222660
https://weekly-net.co.jp/news/37120/
http://www.news24.jp/articles/2018/05/30/07394639.html
https://news.nifty.com/article/item/neta/12141-5431/
http://www.dm-net.co.jp/calendar/2018/028118.php
https://response.jp/article/2018/05/31/310357.html
https://prtimes.jp/main/html/rd/p/000000013.000028314.html
http://www.trt.net.tr/japanese/jian-kang-sheng-huo/2018/06/05/shi-jie-nocheng-ren-no3ren-ni1ren-gashui-mian-bu-zu-985932
https://www.bcnretail.com/market/detail/20180603_62819.html
http://www.sankeibiz.jp/business/news/180607/prl1806071142067-n1.htm
https://www.mylohas.net/2018/06/168913sleep.html
https://toyokeizai.net/articles/-/223607
https://diamond.jp/articles/-/172140
http://news.nicovideo.jp/watch/nw3583626
http://cargo-news.co.jp/cargo-news-main/1009
https://weekly-net.co.jp/news/37490/
https://diamond.jp/articles/-/172262

index.js
'use strict';

const fs = require('fs');
const path = require('path');

const _ = require('lodash');
const request = require('request-promise');

const uri = process.env.SLACK_URI || require('./config').slack.uri;
const { getArticles } = require('./scrape');

try {
  (async () => {
    console.log('Getting articles...');
    const list = await getArticles();
    console.log(list);
    const { href, innerText } = _.sample(list);
    const text = `今日の記事:「${innerText}」\n${href}`;
    console.log(text);
    const opts = {
      method: 'POST',
      uri,
      headers: {
        'Content-Type': 'application/json'
      },
      body: { text },
      json: true
    };
    await request(opts);
    const filepath = path.resolve(__dirname, 'history');
    const file = fs.readFileSync(filepath, 'utf8');
    fs.writeFileSync(filepath, `${file}${href}\n`);
  })();
} catch (e) {
  throw e;
}


scrape.js
'use strict';

const fs = require('fs');
const path = require('path');

const _ = require('lodash');
const Aigle = require('aigle');
const puppeteer = require('puppeteer');

const history = fs.readFileSync(path.resolve(__dirname, 'history'), 'utf8');
const historyMap = _.chain(history)
  .trim()
  .split('\n')
  .transform((map, text) => map[text] = text, {})
  .value();

Object.assign(exports, { getArticles });

const url = 'https://www.google.co.jp/search?tbm=nws&q=%E7%9D%A1%E7%9C%A0%E3%80%80%E3%83%8B%E3%83%A5%E3%83%BC%E3%82%B9';
const words = [
  '不足',
  '病',
  '原因',
  '血圧',
  '死',
  '症',
  '質',
  '負',
  '免疫',
  '悪',
  'ミス'
];
const re = new RegExp(words.join('|'));

async function getArticles() {
  const browser = await puppeteer.launch({ args: ['--no-sandbox', '--disable-setuid-sandbox'] });
  const page = await browser.newPage();

  // set logal storage
  let count = 0;
  const tester = list => !list.length && ++count < 10;
  const iterator = async () => {
    const u = count ? `${url}&start=${count*10}` : url;
    const list = await getResult(u);
    console.log(`page:${count}\n`, _.map(list, 'innerText'));
    return _.filter(list, ({ href, innerText }) => !historyMap[href] && re.test(innerText));
  }
  const result = await Aigle.doWhilst(iterator, tester);
  await browser.close();

  return result;

  async function getResult(url) {
    await page.goto(url);
    return await page.evaluate(() => {
      const links = document.querySelectorAll('.top');
      const texts = document.querySelectorAll('h3');
      const result = [];
      links.forEach(({ href }, i) => result.push({
        href,
        innerText: texts[i].innerText
      }));
      return result;
    });
  }
}
